Cluster features on raft: topology coordinator + check on boot followups

piodul opened the pull request:
This PR collects followups described in #14972:

- The `system.topology` table is now flushed every time feature-related columns are modified. This is done because of the feature check that happens before the schema commitlog is replayed.
- The implementation now guarantees that, if all nodes support some feature as described by the `supported_features` column, then support for that feature will not be revoked by any node.
  Previously, in an edge case where a node is the last one to add support for some feature `X` in `supported_features` column, crashes before applying/persisting it and then restarts without supporting `X`, it would be allowed to boot anyway and would revoke support for the `X` in `system.topology`.
  The existing behavior, although counterintuitive, was safe - the topology coordinator is responsible for explicitly marking features as enabled, and in order to enable a feature it needs to perform a special kind of a global barrier (`barrier_after_feature_update`) which only succeeds after the node has updated its features column - so there is no risk of enabling an unsupported feature.
  In order to make the behavior less confusing, the node now will perform a second check when it tries to update its `supported_features` column in `system.topology`.
- The `barrier_after_feature_update` is removed and the regular global `barrier` topology command is used instead. The `barrier` handler now performs a feature check if the node did not have a chance to verify and update its cluster features for the second time.

JOIN_NODE rpc will be sent separately as it is a big item on its own.

Fixes: #14972

piodul commented on service/storage_service.cc:
```diff
@@ -5678,6 +5678,28 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                    // TODO: Somehow make it stop the boot process if it fails
+                    const auto& enabled_features = _topology_state_machine._topology.features.enabled_features;
+                    const auto unsafe_to_disable_features = _topology_state_machine._topology.features.calculate_not_yet_enabled_features();
+                    _feature_service.check_features(enabled_features, unsafe_to_disable_features);
```
@gleb-cloudius in our discussion under the previous PR (https://github.com/scylladb/scylladb/pull/14722#discussion_r1285522121) you wrote that the feature check that happens in the `barrier` command handler should cause the node to shut down, however I have no idea how to propagate the failure from here in a sensible way to cause a shutdown. Currently, the startup procedure will eventually reach `check_features` in `update_topology_with_local_metadata` and this will shut down the node.

I'll be happy to implement it if you have a good suggestion on how to do it.

gleb-cloudius replied:
We can print an appropriate error message an call ss::drain().

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3270/

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -5678,6 +5678,28 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                    // TODO: Somehow make it stop the boot process if it fails
+                    const auto& enabled_features = _topology_state_machine._topology.features.enabled_features;
+                    const auto unsafe_to_disable_features = _topology_state_machine._topology.features.calculate_not_yet_enabled_features();
+                    _feature_service.check_features(enabled_features, unsafe_to_disable_features);
```
It should be enough to check that there is no feature in  `_topology_state_machine._topology.features.enabled_features` that are also not in  `_feature_service.supported_feature_set()`. We can also do it unconditionally. This is not a fast path.

piodul replied:
I agree with what you wrote, but instead of `enabled_features` we should use `unsafe_to_disable_features`.

piodul replied:
Actually, I'd like to stick to `check_features` due to the message it produces via exception - some tests expect this message to be printed into logs.

gleb-cloudius replied:
I would like to get rid of `_topology_updated_with_local_metadata`. Simple check can be done unconditionally. I wouldn't want tests dictate the code. Can we print the same message if the check fails?

If we cannot get rid of `_topology_updated_with_local_metadata` then it should not be a bool but a feature that is waited here instead of the check.

piodul replied:
I removed `_topology_updated_with_local_metadata`.

I agree that the tests should not dictate the code, but I'd also like to keep the error messages to be the same for the sake of users reading the logs (imagine that you get a different message 1% of the times - weird and confusing). The exception message is also slightly different depending on the circumstances, so I resorted to catching the exception and printing its message in the logs.

gleb-cloudius replied:
I actually usually prefer to have slightly different messages. I know what code path failed then. It can be just a message prefix like the function name.

piodul replied:
Hm, alright, actually the message is now slightly different that what is printed in the usual case - it will be prefixed with `raft topology: feature check during barrier failed: `, but the part that follows (extracted from the exception) is the same.

gleb-cloudius replied:
Perfect.

piodul commented:
v2:
- Moved the "Enabling persisted feature" message from `check_features` to the more fitting `enable_features_on_startup`
- Added `unsupported_feature_exception`
- Changed the `barrier` command to call `storage_service::drain()` in the corner case where the node does not support a feature previously advertised by every node

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3331/

piodul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3331/

The failures seem unrelated to the PR:

> Tests / Sanity Tests / sstable_generation_loading_test.TestSSTableGenerationAndLoading.test_sstableloader_case_sensitive_with_quotas

The error message is different, but maybe it is related to https://github.com/scylladb/scylladb/issues/15210 ? @tchaikov

> Tests / Unit Tests / test_fencing.test_fence_hints
> Tests / Unit Tests / non-boost tests.topology_experimental_raft.test_fencing.debug.1

Looks like the fencing test for hints failed. @gusev-p

tchaikov commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3331/
> 
> The failures seem unrelated to the PR:
> 
> > Tests / Sanity Tests / sstable_generation_loading_test.TestSSTableGenerationAndLoading.test_sstableloader_case_sensitive_with_quotas
> 
> The error message is different, but maybe it is related to #15210 ? @tchaikov

it's more like https://github.com/scylladb/scylla-dtest/issues/3464. looking

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2259,6 +2259,20 @@ future<> storage_service::do_update_topology_with_local_metadata(raft::server& r
+        // Fortunately, there is no risk that this feature was marked as enabled
+        // because it requires that the current node responded to a barrier
+        // request - which will fail in this situation.
+        const auto& enabled_features = _topology_state_machine._topology.features.enabled_features;
```
This shouldn't be the features from the topology because we may not wrote them yet. It should be features provided on the command line.

piodul replied:
I don't understand. `check_features` accepts two arguments: a list of features considered enabled by the cluster, and a list of features unsafe to disable, according to the cluster (i.e. features supported by everybody but not yet enabled).

Why would we pass supported features as the first argument? If you look at the `check_features` implementation, it calculates the `supported_feature_set()` based on the node's configuration/CLI arguments and checks them against arguments, so this looks correct to me.

gleb-cloudius replied:
OK. I misunderstood what `check_features` receives as a parameters. I thought the first one is currently supported feature by this node. But then why have two parameters to the function? All features in both sets should be present so why not just merge them before passing to the function? Is it just to get different error message?

piodul replied:
Yes, the function produces a different message depending on whether a feature is missing from the first or the second set. I can see why it looks confusing, though I don't have an idea how to make it look better.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -5677,9 +5677,39 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                std::exception_ptr ex;
+                try {
+                    const auto& enabled_features = _topology_state_machine._topology.features.enabled_features;
+                    const auto unsafe_to_disable_features = _topology_state_machine._topology.features.calculate_not_yet_enabled_features();
```
Can this be precalculated when we load the topology?

piodul replied:
Yes, but is it worth it? It would have to be recalculated every time we call `system_keyspace::load_topology_state`.

gleb-cloudius replied:
Not sure. When `system_keyspace::load_topology_state` will become smarter and will amend the in memory topology structures instead of re-creating them it will be worth it for sure.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3382/
